### PR TITLE
feat(oas) - accurate model OAS representation - PA to PU

### PR DESCRIPTION
**Pull request by patrick-medusajs**
### Scope

Models PA to PU

### What

Refactor OAS for models to accurately represent their shape in API responses.

### Why

About 33% of model fields are not accurately represented in the OAS. Most of the issues are:
- fields that can not be omitted in the response are not declared as `required`
- fields that could return `null` as their value are not declared as `nullable: true`

When using a code generator, these OAS issues would lead to inaccurate response shapes in the generated client.

#### nullable
Fields meeting at least one of the following condition will be represented as `nullable: true` in OAS:
* The field is decorated with `@Column({ nullable: true })`
* The field is decorated with `@OneToOne`, `@ManyToOne`
* The field is decorated with `@DeleteDateColumn`

#### optional
Fields meeting at least one of the following conditions will never be listed as `required` in OAS and will be considered optional and could be omitted in the response:
* The field is decorated with `@OneToOne`, `@ManyToOne`, `@OneToMany`, `@ManyToMany`
* The field is decorated with `@FeatureFlagColumn`
* The field is decorated with `@Column({select: false})`
* The field is representing dynamic values not persisted in the database

Fields not meeting any of the conditions above will be declared as `required` and are expected to be present in the response.

### Test
* Ran OAS validator.
* Ran docs build script.

Expect OAS changes to be reflected in the API documentation.

### CODE REVIEW

**Review by patrick-medusajs — COMMENTED** [review]
self-review: self-review: I'd recommend using the [Model TS refactor PR](https://github.com/medusajs/medusa/pull/3202/files) to facilitate the review process. I've also create a [demo PR](https://github.com/medusajs/medusa/pull/3204/files) to showcase the generate types from the OAS.

Note that `product-variant` and `product-variant-inventory-item` have changed since I created the TS and Demo PR.

UPDATE: I just updated the [TS PR]((https://github.com/medusajs/medusa/pull/3202/files))

File: packages/medusa/src/models/product.ts [hunk 1]
```diff
@@ -248,98 +275,118 @@ export class Product extends SoftDeletableEntity {
  *     type: array
  *     items:
  *       $ref: "#/components/schemas/ProductVariant"
+ *   categories:
+ *     description: Available if the relation `category_children` are expanded.
+ *     type: array
+ *     items:
+ *       $ref: "#/components/schemas/ProductCategory"
```
Comment (shahednasser) [inline packages/medusa/src/models/product.ts:282]: `categories` is available twice (an old definition is available below)

**Review by shahednasser — CHANGES_REQUESTED** [review]

File: packages/medusa/src/models/product-category.ts [hunk 2]
```diff
@@ -114,31 +123,38 @@ export class ProductCategory extends SoftDeletableEntity {
...
  *     description: A product category object. Available if the relation `parent_category` is expanded.
- *     type: object
+ *     nullable: true
+ *     $ref: "#/components/schemas/ProductCategory"
+ *   products:
+ *     description: Available if the relation `products` is expanded.
+ *     type: array
+ *     items:
+ *       $ref: "#/components/schemas/Product"
  *   products:
```
[the remark this answers is not in the document]
Comment (patrick-medusajs) [reply packages/medusa/src/models/product-category.ts]: 😅 Fixed 4203dc61870fc6042e2acef3beb0481f5e3d2aa6

File: packages/medusa/src/models/product-variant-inventory-item.ts [hunk 3]
```diff
@@ -38,31 +38,43 @@ export class ProductVariantInventoryItem extends SoftDeletableEntity {
...
+ *     description: The date with timezone at which the resource was created.
  *     type: string
- *     description: "The date with timezone at which the resource was created."
  *     format: date-time
  *   updated_at:
+ *     description: The date with timezone at which the resource was updated.
  *     type: string
- *     description: "The date with timezone at which the resource was updated."
  *     format: date-time
  *   deleted_at:
```
[the remark this answers is not in the document]
Comment (patrick-medusajs) [reply packages/medusa/src/models/product-variant-inventory-item.ts:76]: Fixed in 4203dc61870fc6042e2acef3beb0481f5e3d2aa6

File: packages/medusa/src/models/product.ts (hunk 1 above)
Comment (patrick-medusajs) [reply packages/medusa/src/models/product.ts]: 😅 Fixed in 4203dc61870fc6042e2acef3beb0481f5e3d2aa6

**Review by shahednasser — APPROVED** [review]
LGTM

**Pull request merged**